### PR TITLE
Enhanced exception handling

### PR DESCRIPTION
**Pull request by ob-ab**
- Handled Different Types of Exceptions with Multiple Catch Blocks
- Added Throw Command to throw Custom Exception
- Throwing Exception out of Catch block and passing from Child Script to Parent (Calling) Script
- Added Rethrow Command to throw the original exception
- Added Retry Command to retry the contained commands as long as the condition is not met or an error is thrown.

### CODE REVIEW

File: taskt/Core/Automation/Commands/Error Handling Commands/CatchCommand.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        [PropertyUISelectionOption("InvalidDataException")]
+        [PropertyUISelectionOption("InvalidOperationException")]
+        [PropertyUISelectionOption("KeyNotFoundException")]
+        [PropertyUISelectionOption("NotSupportedException")]
+        [PropertyUISelectionOption("NullReferenceException")]
+        [PropertyUISelectionOption("OverflowException")]
+        [PropertyUISelectionOption("TimeoutException")]
+        [InputSpecification("Select the appropriate exception type.")]
+        [SampleUsage("")]
+        [Remarks("If the type of the exception occurred in the try block, matches the selected exception type then this command will be executed.")]
```
Comment (openbots-ff) [inline taskt/Core/Automation/Commands/Error Handling Commands/CatchCommand.cs]:
```suggestion
        [Remarks("This command will be executed if the type of the exception that occurred in the try block matches the selected exception type.")]
```

File: taskt/Core/Automation/Commands/Error Handling Commands/BeginRetryCommand.cs [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+            _ifConditionHelper.AllowUserToAddRows = false;
+            _ifConditionHelper.AllowUserToDeleteRows = true;
+            _ifConditionHelper.CellContentClick += IfConditionHelper_CellContentClick;
+
+            return RenderedControls;
+        }
+
+        public override string GetDisplayValue()
+        {
+            return base.GetDisplayValue();
```
Comment (openbots-ff) [inline taskt/Core/Automation/Commands/Error Handling Commands/BeginRetryCommand.cs]: Requires additional context

File: taskt/Core/Automation/Commands/Error Handling Commands/ThrowCommand.cs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            base.Render(editor);
+
+            RenderedControls.AddRange(CommandControls.CreateDefaultDropdownGroupFor("v_ExceptionType", this, editor));
+            RenderedControls.AddRange(CommandControls.CreateDefaultInputGroupFor("v_ExceptionMessage", this, editor));
+            return RenderedControls;
+        }
+
+        public override string GetDisplayValue()
+        {
+            return base.GetDisplayValue() + $" '{v_ExceptionType}'";
```
Comment (openbots-ff) [inline taskt/Core/Automation/Commands/Error Handling Commands/ThrowCommand.cs]: Maybe add the message as well

File: taskt/Core/Automation/Commands/Error Handling Commands/RethrowCommand.cs [hunk 4]
```diff
@@ -9,24 +9,20 @@ namespace taskt.Core.Automation.Commands
...
+            CommandName = "RethrowCommand";
+            SelectionName = "Rethrow";
             CommandEnabled = true;
             CustomRendering = true;
         }
 
         public override void RunCommand(object sender)
         {
-            //no execution required, used as a marker by the Automation Engine
+            throw new Exception("Rethrowing Original Exception");
```
Comment (openbots-ff) [inline taskt/Core/Automation/Commands/Error Handling Commands/RethrowCommand.cs:25]: You're not really rethrowing the original exception though. This is a brand new one. None of the original exception data will be passed on with this command. Or am I missing something?

File: taskt/Core/Automation/Commands/Error Handling Commands/BeginRetryCommand.cs [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        [PropertyUIHelper(PropertyUIHelper.UIAdditionalHelperType.ShowVariableHelper)]
+        public string v_RetryInterval { get; set; }
+
+        [XmlElement]
+        [PropertyDescription("Condition")]
+        [InputSpecification("Add a condition.")]
+        [SampleUsage("")]
+        [Remarks("Items in the retry scope will be executed, if the condition doesn't satisfy.")]
+        [PropertyUIHelper(PropertyUIHelper.UIAdditionalHelperType.ShowIfBuilder)]
+        public DataTable v_IfConditionsTable { get; set; }
```
Comment (openbots-ff) [inline taskt/Core/Automation/Commands/Error Handling Commands/BeginRetryCommand.cs:47]: Was it your intention to disable the ability to add multiple if conditions despite using a ShowIfBuilder with a table? Wouldn't it be better to just replicate the drop down that BeginIf uses instead? I think you should either make it possible to add multiple conditions or remove the table entirely and replace it with a drop down if you're just going to use one.

File: taskt/Core/Automation/Commands/Error Handling Commands/BeginRetryCommand.cs [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        [SampleUsage("5 || {vRetryInterval}")]
+        [Remarks("")]
+        [PropertyUIHelper(PropertyUIHelper.UIAdditionalHelperType.ShowVariableHelper)]
+        public string v_RetryInterval { get; set; }
+
+        [XmlElement]
+        [PropertyDescription("Condition")]
+        [InputSpecification("Add a condition.")]
+        [SampleUsage("")]
+        [Remarks("Items in the retry scope will be executed, if the condition doesn't satisfy.")]
```
Comment (openbots-ff) [inline taskt/Core/Automation/Commands/Error Handling Commands/BeginRetryCommand.cs]:
```suggestion
        [Remarks("Items in the retry scope will be executed if the condition doesn't satisfy.")]
```

**Review by openbots-ff — CHANGES_REQUESTED** [review]

File: taskt/Core/Automation/Commands/Error Handling Commands/RethrowCommand.cs (hunk 4 above)
Comment (ob-ab) [reply taskt/Core/Automation/Commands/Error Handling Commands/RethrowCommand.cs:25]: This was done intentionally to just throw a generic exception from Rethrow Command to act accordingly in the Catch block of ExecuteCommand Method in AutomationEngineInstance class

File: taskt/Core/Automation/Commands/Error Handling Commands/BeginRetryCommand.cs (hunk 5 above)
Comment (CameronHerwig) [reply taskt/Core/Automation/Commands/Error Handling Commands/BeginRetryCommand.cs:47]: Agreed. I assumed given the UI I could add multiple conditions. A single condition should be sufficient for now.

Comment (ob-ab) [reply taskt/Core/Automation/Commands/Error Handling Commands/BeginRetryCommand.cs:47]: Added ability to add multiple if conditions.

File: taskt/Core/Automation/Commands/Error Handling Commands/ThrowCommand.cs (hunk 3 above)
Comment (ob-ab) [reply taskt/Core/Automation/Commands/Error Handling Commands/ThrowCommand.cs]: done

File: taskt/Core/Automation/Commands/Error Handling Commands/BeginRetryCommand.cs (hunk 2 above)
Comment (ob-ab) [reply taskt/Core/Automation/Commands/Error Handling Commands/BeginRetryCommand.cs]: done

**Review by openbots-ff — APPROVED** [review]

**Review by CameronHerwig — APPROVED** [review]
Looks good. I have a few things to clarify tomorrow, but nothing I'm willing to hold up the PR for. Great work.

**Pull request merged**